modules/light-clients/06-solomachine/types/update.go

```diff
@@ -17,61 +17,79 @@ import (
 // - the currently registered public key did not provide the update signature
 func (cs ClientState) CheckHeaderAndUpdateState(
 	ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore,
-	header exported.Header,
+	msg exported.Header, // TODO: Update to exported.ClientMessage
 ) (exported.ClientState, exported.ConsensusState, error) {
-	smHeader, ok := header.(*Header)
+	if err := cs.VerifyClientMessage(cdc, msg); err != nil {
+		return nil, nil, err
+	}
+
+	// TODO: Remove this type assertion, replace with misbehaviour checking and update state
+	smHeader, ok := msg.(*Header)
 	if !ok {
 		return nil, nil, sdkerrors.Wrapf(
-			clienttypes.ErrInvalidHeader, "header type %T, expected  %T", header, &Header{},
+			clienttypes.ErrInvalidHeader, "expected %T, got %T", &Header{}, msg,
 		)
 	}
 
-	if err := checkHeader(cdc, &cs, smHeader); err != nil {
-		return nil, nil, err
-	}
-
 	clientState, consensusState := update(&cs, smHeader)
 	return clientState, consensusState, nil
 }
 
-// checkHeader checks if the Solo Machine update signature is valid.
-func checkHeader(cdc codec.BinaryCodec, clientState *ClientState, header *Header) error {
-	// assert update sequence is current sequence
-	if header.Sequence != clientState.Sequence {
-		return sdkerrors.Wrapf(
-			clienttypes.ErrInvalidHeader,
-			"header sequence does not match the client state sequence (%d != %d)", header.Sequence, clientState.Sequence,
-		)
-	}
+// VerifyClientMessage checks if the Solo Machine update signature(s) is valid.
+func (cs ClientState) VerifyClientMessage(cdc codec.BinaryCodec, clientMsg exported.Header) error {
+	switch msg := clientMsg.(type) {
+	case *Header:
```

Comment:
we could add a `verifyHeader` function so the switch case doesn't get overloaded

Reply:
Added `verifyHeader` and `verifyMisbehaviour` to split the type switch logic out. I also updated the function signature to include context and store (as I think they're required for 07-tendermint).